[ENH] code sktime/forecasting/gluonts_adapters.py

rajatkriplani opened the pull request:
#### Reference Issues/PRs
Fixes #6800 
#### What does this implement/fix? 
This PR introduces modular improvements and refactoring for the `gluonts_adapters.py` module in `sktime`. It addresses repetitive method overrides, improves modularity, and integrates the newly implemented conversion functions.  
Key changes include:
- **Refactored Base Adapter**:
  - Centralized data conversion logic into reusable methods (`_convert_to_gluonts_format`, `_convert_to_sktime_format`).
  - Modularized model initialization by introducing `_init_model` and enabling dynamic model class specification.
- **DeepAR and N-BEATS Adapters**:
  - Simplified implementation by leveraging the refactored base adapter.
  - Eliminated duplication of trainer initialization and prediction logic.
- **Integration of New Conversion Utilities**:
  - Used `convert_gluonts_result_to_multiindex` and `convert_from_multiindex_to_listdataset` for improved compatibility with `sktime` data formats.

These changes streamline the codebase, reduce redundancy, and make it easier to add new GluonTS-based models in the future.
#### Does your contribution introduce a new dependency? If yes, which one?
No
#### What should a reviewer concentrate their feedback on?
Please focus on:
- The correctness and clarity of the refactored modular design.
- Compatibility with existing `sktime` functionalities and GluonTS models.
- Adequacy of the integration of new conversion functions.
- Potential edge cases or scenarios not covered in the current implementation.

#### Did you add any tests for the change?
No
#### Any other comments?
No
#### PR checklist

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned.
Note: I can't locate `blob/main/.all-contributorsrc`
- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

rajatkriplani commented:
Hi @fkiraly I'm fixing the code for code-quality checks but I'm facing an issue. I imported the following:
```python
from sktime.datatypes._adapter.gluonts_to_pd_multiindex import convert_gluonts_result_to_multiindex
from sktime.datatypes._adapter.pd_multiindex_to_list_dataset import convert_from_multiindex_to_listdataset
from sktime.forecasting.base._base import BaseForecaster
```
but these are not being imported. I tried setting up `PYTHONPATH` by:
```bash
export PYTHONPATH=/home/divya/Code/open_scr:$PYTHONPATH
source ~/.bashrc
```

rajatkriplani commented:
@fkiraly I'm making changes in the code for code-quality checks for that I'm trying to run/test tings locally but I'm unable to import the follwoing:
```python
from sktime.datatypes._adapter.gluonts_to_pd_multiindex import convert_gluonts_result_to_multiindex
from sktime.datatypes._adapter.pd_multiindex_to_list_dataset import convert_from_multiindex_to_listdataset
from sktime.forecasting.base._base import BaseForecaster
```
I even tried setting up the `PYTHONPATH` manually by:
```bash
export PYTHONPATH=[path to sktime repo on my pc]:$PYTHONPATH
source ~/.bashrc
```

note: I've all the requirement fulfilled on the conda environment

yarnabrina commented:
Hi @rajatkriplani , was the issue solved? If not, can you please give a bit more details? For example what was the exact error, what was traceback message, do you have editable install, etc.?

fkiraly reviewed (changes requested):
Tests are failing due to code formatting, see the guide here:
https://www.sktime.net/en/latest/developer_guide/coding_standards.html

rajatkriplani commented:
Hi @yarnabrina sorry for not addressing your comment earlier, I'll ping you in the thread about this issue over discord. And @fkiraly  I've changed the code format, but I'm facing some import issue on my local repo:

![Screenshot from 2025-01-04 01-26-58](https://github.com/user-attachments/assets/960c5feb-941e-4ba6-b538-17a161427d02)
I've made sure to import everything correctly, and uninstalled and re-installed all the required dependencies (dev installation), still facing this issue. Can you tell me what else I can do to fix it and test the code?
below is the test code (yet to commit):
```python
import pytest
import pandas as pd
import numpy as np
from sktime.utils.estimator_checks import check_estimator
from sktime.forecasting.base import ForecastingHorizon
from gluonts_adapters import DeepARAdapter, NBEATSAdapter

# 1. Interface Validation with `check_estimator`
def test_check_estimator():
    """Test interface compliance for DeepARAdapter and NBEATSAdapter."""
    check_estimator(DeepARAdapter)
    check_estimator(NBEATSAdapter)

# 2. Unit Tests for `_fit` and `_predict`
def test_fit_and_predict():
    """Test the _fit and _predict methods."""
    y = pd.Series(np.random.rand(100), index=pd.date_range("2023-01-01", periods=100))
    fh = ForecastingHorizon(np.arange(1, 6), is_relative=True)

    # Test DeepARAdapter
    forecaster = DeepARAdapter(prediction_length=5, epochs=1)
    forecaster.fit(y)
    assert forecaster._is_fitted, "DeepARAdapter is not fitted after calling fit."
    y_pred = forecaster.predict(fh)
    assert isinstance(y_pred, pd.DataFrame), "DeepARAdapter prediction output is not a DataFrame."

    # Test NBEATSAdapter
    forecaster = NBEATSAdapter(prediction_length=5, epochs=1)
    forecaster.fit(y)
    assert forecaster._is_fitted, "NBEATSAdapter is not fitted after calling fit."
    y_pred = forecaster.predict(fh)
    assert isinstance(y_pred, pd.DataFrame), "NBEATSAdapter prediction output is not a DataFrame."
```

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,95 @@
+from sktime.forecasting.base._base import BaseForecaster
+
+
+class GluonTSBaseAdapter(BaseForecaster):
```
Is this class intended to be public to sktime users? If not, I suggest you prefix this with `_` and create within `sktime/forecasting/base/adapters/_gluonts.py` or similarly named module, and then inherit these for concrete estimators in `sktime/forecasting/gluonts.py` or a similarly named module.

If it was intended to be public, then it probably be placed under `sktime/forecasting/adapters`, but @fkiraly can comment on that.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,95 @@
+class GluonTSBaseAdapter(BaseForecaster):
+    """Common adapter base class for GluonTS models."""
+
+    def __init__(self, model_class, freq="D", prediction_length=24, epochs=10, **kwargs):
```
While I'm personally fi e with this format, I believe it's inconsistent with sktime codebase in general and also with sklearn estimator guidelines. The guidelines suggest to not have any unspecified argument in `__init__` methods, and allowing arbitrary `kwargs` break that. See this: https://scikit-learn.org/stable/developers/develop.html#instantiation

(It's probably not the best link, but that's all I could find now.)

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,95 @@
+class DeepARAdapter(GluonTSBaseAdapter):
+    """Adapter for the GluonTS DeepAR model."""
+
+    def __init__(self, **kwargs):
```
Affected by the kwargs comment

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,95 @@
+class NBEATSAdapter(GluonTSBaseAdapter):
+    """Adapter for the GluonTS N-BEATS model."""
+
+    def __init__(self, **kwargs):
```
Affected by the kwargs comment

yarnabrina reviewed (changes requested):
I've made a few comments, kindly take a look. Also, code quality checks are failing, please ensure you are using pre-commit as documented in the contribution guidelines.

The import errors you mentioned is difficult/impossible for me atleast to debug without live checking, so your best bet will probably be to attend one of the Friday connects and do a screen share.

rajatkriplani commented:
Hi @yarnabrina  I made changes according to your feedback 
 - `GluonTSBaseAdapter` renamed to `_GluonTSBaseAdapter`
 - removed  `**kwargs` parameter
 - added `trainer_kwargs` instead (additional parameters for the `Trainer`)
 - docstrings added to each `__init__`
 - classes explicitly define parameters
 - `_is_fitted` flag sets True after training (in `_fit`)

yarnabrina commented:
@rajatkriplani please excuse me if I am missing something obvious, but I am unable to see a single thing you mentioned above. Is there a chance that you not have committed or committed but not pushed?

rajatkriplani commented:
> @rajatkriplani please excuse me if I am missing something obvious, but I am unable to see a single thing you mentioned above. Is there a chance that you not have committed or committed but not pushed?

@yarnabrina  Sorry for the inconvenience. I accidently committed the previous version again. Now it's fixed please have a look and let me know your feedback.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+from typing import Optional, Dict, Any
+
+from sktime.forecasting.base import BaseForecaster
+from gluonts.dataset.common import ListDataset
+from gluonts.mx.trainer import Trainer
```
These are problematic. These imports assume gluonts is already present in user's environment, which may not be the case. And if someone wants to use sktime without using your estimator, they can not as these lines will cause failures.

Please take a look at https://www.sktime.net/en/stable/developer_guide/dependencies.html#handling-soft-dependencies. Essentially you have to remove these from module level imports, and move those to methods of the class where it'll be ensured that method does not run unless the dependencies are installed.

rajatkriplani replied:
You're absolutely right, I'll refactor the code. And I'll go through the doc once I have replied to your feedback.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        freq: str = "D",
+        prediction_length: int = 24,
+        epochs: int = 10,
+        trainer_kwargs: Optional[Dict[str, Any]] = None,
```
Small note about type hint here: we have moved on to python 3.9+ support, and we can use `dict` directly. You need not use `typing.Dict` any longer.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+
+    def __init__(
+        self,
+        model_class: Any,
```
I am not sure how I feel about this to be honest. You actually do not support anything, it has to be a estimator class from gluonts. But I am unable to think of a better way to type hint here, so let's wait for review from rest of @sktime/core-developers.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        epochs: int = 10,
+        trainer_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Initialize the GluonTS base adapter.
+
+        Parameters
+        ----------
+        model_class : class
+            GluonTS model class to instantiate.
+        freq : str, optional (default="D")
+            Frequency of the time series.
+        prediction_length : int, optional (default=24)
+            Number of time steps to predict.
+        epochs : int, optional (default=10)
+            Number of training epochs.
+        trainer_kwargs : dict, optional (default=None)
+            Additional keyword arguments for the GluonTS Trainer.
+        """
```
It is customary that `__init__` methods do not need docstrings of their own, and they are documented as part of the class docstring itself. You can move this section to the above.

rajatkriplani replied:
I'll consolidate the parameter descriptions into the class-level docstring.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        self.model_class: Any = model_class
+        self.trainer_kwargs: Dict[str, Any] = trainer_kwargs or {}
+
+        self.model = self._init_model()
```
My suggestion will be to not call this private method here. Initiate the `self.model` attribute as `None` here, and then call the method during `_fit`.

rajatkriplani replied:
Working on it along with other stuff.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        self.trainer_kwargs: Dict[str, Any] = trainer_kwargs or {}
+
+        self.model = self._init_model()
+        self._is_fitted: bool = False
```
Why do you need to define this? It is already part of `BaseForecaster`, any reason to override?

rajatkriplani replied:
Yes thanks for points that out, I will make sure to above redundancy in code from now on.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        self.prediction_length: int = prediction_length
+        self.epochs: int = epochs
+        self.model_class: Any = model_class
+        self.trainer_kwargs: Dict[str, Any] = trainer_kwargs or {}
```
Please do not do this, as it breaks the sktime convention (or possibly in general sklearn convention) that user passed arguments are available as it is as instance properties. You should keep `self.trainer_kwargs` as `trainer_kwargs` only, but initiate a new private property `self._trainer_kwargs` with the `None` handling.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        )
+        return model
+
+    def _convert_to_gluonts_format(self, y: pd.Series) -> ListDataset:
+        """
+        Convert sktime data to GluonTS ListDataset format.
+
+        Parameters
+        ----------
+        y : pd.Series
+            The time series data to convert.
+
+        Returns
+        -------
+        gluonts_dataset : ListDataset
+            The converted GluonTS ListDataset.
+        """
+        return convert_from_multiindex_to_listdataset(y)
```
Why do we need this method, and not call the imported function directly?

rajatkriplani replied:
Now that I think of it this is just another failed attempt to solve the import issue. I hope you are available for live check-up tomorrow after the meet-up.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        """
+        return convert_from_multiindex_to_listdataset(y)
+
+    def _convert_to_sktime_format(self, gluonts_result: Any) -> pd.DataFrame:
+        """
+        Convert GluonTS forecast result to sktime format.
+
+        Parameters
+        ----------
+        gluonts_result : Any
+            The forecast result from GluonTS.
+
+        Returns
+        -------
+        sktime_forecast : pd.DataFrame
+            The converted forecast in sktime's MultiIndex format.
+        """
+        return convert_gluonts_result_to_multiindex(gluonts_result)
```
Same confusion as above.

rajatkriplani replied:
Same import issue.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        """
+        return convert_gluonts_result_to_multiindex(gluonts_result)
+
+    def _fit(self, y: pd.Series, X: Optional[Any] = None, fh: Optional[Any] = None) -> "_GluonTSBaseAdapter":
```
If you are into type hints, why skip `self`? :)

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        """
+        gluonts_data = self._convert_to_gluonts_format(y)
+        self.model.train(gluonts_data)
+        self._is_fitted = True
```
I'm quite certain that this is unnecessary, as it is taken care by `BaseForecaster`.

yarnabrina commented on sktime/forecasting/gluonts_adapters.py:
```diff
@@ -0,0 +1,259 @@
+        """
+        gluonts_data = self._convert_to_gluonts_format(self._y)
+        forecast_it, _ = self.model.make_evaluation_predictions(
+            gluonts_data, num_samples=100
```
Should we let user specify `num_samples` themselves?

rajatkriplani replied:
Oh yes for more flexibility, thanks once again.

yarnabrina reviewed (changes requested):
I've made few comments, and these are not super exhaustive. Please take a look, and let me know if it makes sense to you and/or you need any help from us.

Also, please take a look at https://www.sktime.net/en/stable/developer_guide/coding_standards.html.

OmBiradar commented:
Can I work on this @rajatkriplani @yarnabrina if no one is currently working on this?
